Use modern (target based) CMake

JonasProgrammer opened the pull request:
Hi,

as ozo is [currently switching](https://github.com/yandex/ozo/pull/143) to a more modern CMake approach, I figured perhaps this could be useful for `resource_pool` as well.

I've ported some changes I made to the submodule to the standalone `resource_pool` repository. Basically, the idea is to enjoy the benefits of modern CMake, such as `target_compile_features` while still maintaining compatibility for the packaged version (the files generated for find_package should be compatible with CMake 2.5ish).
The only thing is, that this needs CMake version 3.12 or higher to build; I have no idea if that's a problem for you.

I tried to update the header-only library itself as well as the examples, while keeping the benchmarks and tests the way they are.

Additionally, I've added a recipe for a conan package as well as a test package and integrated them both into the CI. While the latter is not so pretty (the before_script looks a bit scary now), I think the easier portability and integration in new projects is worthwhile.
If desired, you can also extend the travis script to automatically uplopad the package to an approriate conan remote.

Little side-note, I've put the CMake targets into the `yandex::` namespace, mostly because that's what I did when changing stuff in ozo's contrib/ directly. If you want this to have another prefix, such as elsid or yamail, just let me know, I will change the PR accordingly.

JonasProgrammer committed:
Modern CMake
- switched to target based CMake
- use target-based cmake for exmaples
- tests & benchmarks still use legacy CMake, but import the targets
- added proper CPack configuration files
- added Conan recipe

elsid commented on .travis.yml:
```diff
@@ -26,28 +26,39 @@ env:
+- '[ -z "$BOOST_MINOR" ] || ./bootstrap.sh --with-libraries=system,thread,context,coroutine,atomic,date_time'
+- '[ -z "$BOOST_MINOR" ] || mkdir build'
+- '[ -z "$BOOST_MINOR" ] || ./b2 --build-dir=build'
+- '[ -z "$BOOST_MINOR" ] || sudo ./b2 --build-dir=build install'
```
This part could be put into separate .sh file with only single `if`. It would be easier to read.

elsid commented on include/CMakeLists.txt:
```diff
@@ -1 +1,43 @@
+
+export(EXPORT resource_poolTargets
+        FILE "${CMAKE_CURRENT_BINARY_DIR}/resource_pool/resource_poolTargets.cmake"
+        NAMESPACE yandex::)
```
Using yandex namespace is not very appropriate. Change to elsid please.

JonasProgrammer replied:
Sorry, I didn't mean to insult you or discredit your work. As stated, this work started in the depths of ozo'z `contrib/`, so I just copied what I had there.
I have changed it now to properly reflect ownership.

elsid replied:
No problem, I wasn't insulted. It's moslty not about the ownership, but just about where this repository is located.

elsid commented on scripts/travis/conan.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
```
Why sh and not bash? Is there any restrictions to use bash?

JonasProgrammer replied:
I tend to use sh, when bash isn't specifically necessary. I've changed it nontheless to be a little more consistent

elsid commented on scripts/travis/conan.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+set -e
```
Just use shebang: `#!/bin/bash -e` or `#!/bin/sh -e`.

elsid commented on conanfile.py:
```diff
@@ -0,0 +1,48 @@
+
+    # Internal option for ozo devs to use cmake for easier setup
+    options = {'ozo_require_gtest': [True, False]}
+    default_options = {'ozo_require_gtest': False}
```
Why do you mention ozo here and why this option is not used anywhere else? I'm not familiar with conan. Is this option used to build ozo? But it's not mentioned in https://github.com/yandex/ozo/pull/143 . I don't quite understand what is logic behind it.

JonasProgrammer replied:
There is no logic behind that, it's overly enthusiastic copy&paste. I originally didn't have a separate conanfile for `resource_pool` and just had the ozo one pack everything.
For this PR, I figured a separate package is better, as the library is very useful in standalone scenarios. I just copied it over, changed the names and symbols and removed everything ozo-specific -- or so I thought.

As this didn't impact the build process, I simply didn't notice it. Sorry for the inconvenience.

elsid commented on .travis.yml:
```diff
@@ -26,28 +26,39 @@ env:
+- '[ -z "$BOOST_MINOR" ] || wget -O "boost_1_${BOOST_MINOR}_0.tar.gz" "https://dl.bintray.com/boostorg/release/1.${BOOST_MINOR}.0/source/boost_1_${BOOST_MINOR}_0.tar.gz"'
+- '[ -z "$BOOST_MINOR" ] || tar xzf "boost_1_${BOOST_MINOR}_0.tar.gz"'
+- '[ -z "$BOOST_MINOR" ] || cd "boost_1_${BOOST_MINOR}_0"'
+- '[ -z "$BOOST_MINOR" ] || ./bootstrap.sh --with-libraries=system,thread,context,coroutine,atomic,date_time'
```
Why do we need atomic and date_time libraries?

JonasProgrammer committed:
Requested changes for PR
- use bash instead of sh
- use -e option for shell rather than explicit set -e
- extracted boost installation into a separate script
- use elsid namespace
- removed conanfile contents accidentally c&ped from ozo

JonasProgrammer committed:
Use appropriate conan repo

JonasProgrammer commented:
I forgot to resolve the conversations. In theory, everything should be as requested now.
